_posts/2018-02-12-continuous-releases.md

```diff
@@ -0,0 +1,107 @@
+---
+layout: post
+tags: development-process
+date: 2018-02-12 13:14
+title: The Hypothesis continuous release process
+published: true
+author: alexwlchan
+---
+
+If you watch [the Hypothesis changelog][changelog], you'll notice the rate of releases sped up dramatically in 2017.
+We released over a hundred different versions, sometimes multiple times a day.
+
+This is all thanks to our continuous release process.
+We've completely automated the process of releasing, so every pull request that changes code gets a new release, without any human input.
+In this post, I'll explain how our continuous releases work, and why we find it so useful.
+
+[changelog]: https://hypothesis.readthedocs.io/en/latest/changes.html
+
+<!--more-->
+
+## How it works
+
+In the past, Hypothesis was released manually.
+Somebody had to write a changelog, tag a new release on GitHub, and run some manual pip commands to publish a new version to PyPI -- and only David had the credentials for the latter.
+
+This meant that releases were somewhat infrequent, and features spent a long time in master before they were available to `pip install`.
+The pace of development picked up in 2017 -- partly as new maintainers arrived, and partly work for David's PhD -- and we wanted to be able to release more frequently.
+We decided to automate the entire release process.
+
+Now, when you create a pull request that changes the Hypothesis code -- anything that gets installed by pip, not the tests or docs -- you have to include a `RELEASE.rst` file, which describes your change.
+Here's an example from [a recent pull request][recent]:
+
+    RELEASE_TYPE: patch
+
+    This release changes the way in which Hypothesis tries to shrink the size of
+    examples. It probably won't have much impact, but might make shrinking faster
+    in some cases. It is unlikely but not impossible that it will change the
+    resulting examples.
+
+The first line says whether this is a major, minor, or patch release (using [semantic versioning][semver]).
+The rest is a description of the changes in your patch.
+
+We have a test in CI that checks for this file -- any change to the core code needs a release file, even [fixing a typo][typo].
+If you need a release file but haven't written one, the tests fail and your pull request won't be merged.
```

Comment:
Maybe worth mentioning that you can create a `RELEASE.rst` without source changes and we sometimes do for e.g. making `setup.py` changes (which maybe we should just add to the checks TBH) and test changes for the benefit of downstream packagers?